Option resolution with proxies

kurkle opened the pull request:
This is essentially combining best parts of #8008 and #8260 and adding quite a lot more.

There are more places to utilize these superpowers, scales for example.

[A pen to play with](https://codepen.io/kurkle/pen/rNWOdNG)
[Updated pen](https://codepen.io/kurkle/pen/OJbXjPP)

TODO:

- [x] documentation
- [x] scales

kurkle commented:
| Filename | Size | Change |  |
| :--- | :---: | :---: | :---: |
| `dist/chart.esm.js` | 63.3 kB | -368 B (-1%) |  |
| `dist/chart.js` | 80.2 kB | +647 B (+1%) |  |
| `dist/chart.min.js` | 58.2 kB | +400 B (+1%) |  |
| `dist/chunks/helpers.segment.js` | 17.5 kB | +1.17 kB (+7%) | 🔍 |
| `dist/helpers.esm.js` | 1.04 kB | +36 B (+4%) |  |

etimberg reviewed (commented):
Overall, I like the pen and where this is going. I read through all the files and left some comments. I will want to explore this more in detail on the weekend as there are a lot of changes.

kurkle commented on src/core/core.defaults.js:
```diff
@@ -43,15 +48,24 @@ export class Defaults {
 			mode: 'nearest',
 			intersect: true
 		};
+		this.locale = () => navigator ? (navigator.languages ? navigator.languages[0] : navigator.language) : 'en-US';
```
This can also be removed, just something I tested.

etimberg replied:
ok, sounds good

kurkle commented on test/specs/core.controller.tests.js:
```diff
@@ -128,9 +128,9 @@ describe('Chart', function() {
+			expect(options.showLine).toBe(defaults.controllers.line.datasets.showLine);
 			expect(options.spanGaps).toBe(true);
-			expect(options.hover.onHover).toBe(callback);
+			// expect(options.hover.onHover).toBe(callback);
```
Forgotten

kurkle commented on src/controllers/controller.bubble.js:
```diff
@@ -168,6 +149,7 @@ BubbleController.defaults = {
 	plugins: {
 		tooltip: {
 			callbacks: {
+				_scriptable: false,
```
Only for that kind of situations: https://github.com/chartjs/Chart.js/blob/650956b2e1ec8a9591135476cc61f64ddaad7f12/test/specs/plugin.legend.tests.js#L711

That is the only test that actually detects it, but it would apply to tooltip callbacks too if the plugin defaults are removed.

kurkle replied:
It might be better to relocate these "descriptors" so they are not removed when defaults are modified and that would remove the need to replicate.

etimberg replied:
Yeah, i can see if causing problems. Not sure where the best place to move it would be.

kurkle replied:
I moved the descriptors to a private symbol in defaults and added that as a final fallback scope for resolution. Thoughts?

etimberg replied:
Good idea, makes sense to me and will prevent errors

kurkle commented on src/helpers/helpers.core.js:
```diff
@@ -309,3 +309,8 @@ export function resolveObjectKey(obj, key) {
 }
+
+
+export const defined = (value) => typeof value !== 'undefined';
```
The good old `resolve` only considers `undefined` as not defined and continues resolving. `null` can be a valid value for some option.

etimberg reviewed (commented):
I like the descriptors solution

etimberg commented:
Ok, so I tested this out. Only thing I noticed is that custom tooltips are not working in the samples.

`Uncaught Error: Recursion detected: custom->custom`

kurkle commented:
> Ok, so I tested this out. Only thing I noticed is that custom tooltips are not working in the samples.
> 
> `Uncaught Error: Recursion detected: custom->custom`

rebased and added custom as not scriptable for tooltip.

![image](https://user-images.githubusercontent.com/27971921/107112658-e51e0b80-6861-11eb-9627-36f31eeaf0f0.png)

kurkle commented:
Should stop exporting the private helpers publicly. But that would mean rollupping the specs again.
Will not try to change that here, because there were some issues previously with it, like non matching the line numbers in karma reports and incorrect test coverage.

kurkle commented:
I've made quite a few adjustments to make the code a bit clearer. Also noticed animated charts were slower, due no caching of animations, so fixed that.

But there are things that are not really related to this pr. I'll pr them separately. Like the codeclimate config changes :)

kurkle commented:
| Filename | Size | Change |  |
| :--- | :---: | :---: | :---: |
| `dist/chart.esm.js` | 63.6 kB | -192 B (0%) |  |
| `dist/chart.js` | 80.8 kB | +1.11 kB (+1%) |  |
| `dist/chart.min.js` | 58.6 kB | +750 B (+1%) |  |
| `dist/chunks/helpers.segment.js` | 17.9 kB | +1.5 kB (+9%) | 🔍 |
| `dist/helpers.esm.js` | 1.05 kB | +44 B (+4%) |  |

etimberg commented on src/core/core.datasetController.js:
```diff
@@ -764,18 +762,13 @@ export default class DatasetController {
 		const scopes = config.getOptionScopes(me.getDataset(), scopeKeys);
 		const names = Object.keys(defaults.elements[elementType]);
+		// context is provided as a function, and is called only if needed,
+		// so we don't crate a context for each element if not needed.
```
[nit] typo in the comment

kurkle commented:
Any thoughts on how to document this?

etimberg commented:
We could make a new `design-docs` top level folder and put some markdown in there? Or we could add it to the regular docs as  pages under the developers section. I think we probably want to document the resolution order and how to control it, how to disable scriptable/index options, etc.

For the user docs, we can probably keep the current format with indexable and scriptable table columns, but we'd need to add it everywhere. Would that be too tedious?

kurkle commented:
I had changed the order of imports in core.datasetController, reverted that and the naming change on `cloneIfNotShared`. Just to make this a tiny bit smaller change.

kurkle commented:
Marked this as ready for review, although the documentation is not very extensive. 
At least the developer docs need to be provided still, but I think it might be ok to do that separately.

benmccann commented on docs/docs/general/options.md:
```diff
@@ -2,9 +2,74 @@
 
+## Option resolution
+
+Options are resolved from top to bottom, using a context dependant route.
```
I think either one is fine, but it looks like "dependant" is the british spelling and "dependent" is the american spelling, so the latter is more common

kurkle replied:
Ok, our schools favored british english, at least in the 2nd millennium when I was learning :)

benmccann commented on docs/docs/general/options.md:
```diff
@@ -2,9 +2,74 @@
+
+* options.plugins[`plugin.id`]
+* options.controllers[`config.type`].plugins[`plugin.id`]
+* (options.[`...plugin.additionalOptionScopes`])
```
It could be useful to clarify what `additionalOptionScopes` is

kurkle replied:
Yeah, the section where that could link to, is still missing :)

benmccann commented:
Is there a runtime performance impact?

kurkle commented:
> Is there a runtime performance impact?

Not in the usual test cases, because the overhead is masked by caching/sharing when possible. Also the overhead of merging is removed, so its as fast in terms I can measure.

etimberg commented on test/fixtures/scale.radialLinear/gridlines-scriptable.js:
```diff
@@ -11,10 +11,10 @@ module.exports = {
 						display: true,
 						color: function(context) {
-							return context.index % 2 === 0 ? 'red' : 'green';
+							return context.index % 2 === 0 ? 'green' : 'red';
```
Why was this change needed?

kurkle replied:
scriptable and indexable had different `index` base, which is not supported anymore

simonbrunel commented:
It's a complex PR and it's hard to get all consequences. @etimberg already approved it so I will not look at it in details.

Here are some feedback / concerns based on what I read from the [datalabels PR](https://github.com/chartjs/chartjs-plugin-datalabels/pull/221):
1. I like the use of proxies to make easier resolving options, however I think it should be only used internally and not leak to the user (e.g. scriptable context). This logic is not trivial and I'm a bit worry to lose control over it if users rely on it.

2. Options should not be implicitly proxified for external code, meaning plugins should not be called with a proxified `options` as hook last parameter. Instead, plugins should opt-in this feature by creating a proxified options via an helpers and store it internally. This would simplify the plugin API by decoupling it from the options logic. And in some cases (e.g. datalabels), plugins may not want to use this proxy logic at all when it doesn't fit there design.

3. I'm not convinced about mixing naming convention for the `_` prefix. This has always been reserved for private members and we should not require plugin developers to eventually hit these properties (e.g. [`_fallback, _scriptable`](https://github.com/chartjs/chartjs-plugin-datalabels/pull/221/files#diff-598b75d6dc772eb13aa7141f35bb93046ae26c58a97e8647ecf7f74aec8ae93fR150)). Instead, we should move child properties under it's own namespace:

```js
descriptors: {
  fallback: 'foo',
  scriptable: ...,
  children: { // or 'options' or 'properties' or ...
    labels: {
        fallback: 'bar',
        //...
    }
  }
}
```
```

etimberg commented:
For `_fallback` and `_scriptable`, could we use `$fallback` and `$scriptable` instead? A different descriptor format also works, but I'm not sure how painful that is to change.

If plugins didn't receive the proxied objects, would they just receive the merged options as before? That might be a maintenance burden to have essentially two distinct option lookup methods and I could see plugins accidentally breaking in the future without a lot of testing.

simonbrunel commented:
>  could we use `$fallback` and `$scriptable` instead?

I would also consider `$` [a special case](https://chartjs-plugin-datalabels.netlify.app/guide/). But should we really forbid plugins to use `_` or `$` for there option names? Adding an explicit namespace for child properties doesn't look that bad and may help to keep the API consistent and future proof.

>  That might be a maintenance burden to have essentially two distinct option lookup methods

Good point. So if plugins receive the proxied options, at least we should not assume how the plugin wants to fallback. It should be fully under the plugin control and *explicitly* defined by itself (thus not implicitly fallback to the root options / defaults). 

For example, we should not assume that `options.plugins.foo.color` wants to fallback to `options.color` or `defaults.color`. It's not because they have the same name that they relate to each other. Also it adds more implicit / magic logic to maintain while it could be confusing for plugin developers where the values come from.

So I think plugin proxied options should fallback that way:

- options.plugins[`plugin.id`]
- options.controllers[`config.type`].plugins[`plugin.id`]
- defaults.controllers[`config.type`].plugins[`plugin.id`]
- defaults.plugins[`plugin.id`]

Same as with the current merge strategy.

etimberg commented:
I agree that in general, we shouldn't merge the colors for plugins. The default plugins do rely on that behaviour though, so having a per-plugin way to define the fallback seems like a good way to support both of those.

kurkle commented:
> But should we really forbid plugins to use `_` or `$` for there option names? 

No, we'd not forbid `_` or `$`. we'd just resolve `$scritable`, `$indexable` and `$fallback` during option resolution. So those are special. I initally included those in `defaults`, but its common practicde to set the whole `defaults` object of a plugin for eaxmple and we don't want these to be "lost". So the descriptors is currently actually just another fallback scope for defaults, just one that is not overridable by user.

Another option would be to define symbols for those, that would make sure not to collide with anything a plugin developer decides to use. It would also be somewhat akward to use, a plugin would need to import those symbols to use them.

But I will take a look how much it would require duplication to resolve those completely separately.

> It should be fully under the plugin control and _explicitly_ defined by itself (thus not implicitly fallback to the root options / defaults).

Sure, the fallback to root scope can be a additional scope for plugins that want to be simple to use :)

etimberg commented on src/plugins/plugin.legend.js:
```diff
@@ -621,5 +621,7 @@ export default {
-  }
+  },
+
+  additionalOptionScopes: ['']
```
Is it worth a comment saying why we do this?
